algolia/algoliasearch-sfcc-b2c
# feat: color variation swatches

# Color variations swatches support

![image](https://github.com/algolia/algoliasearch-sfcc-b2c/assets/2042109/7e4c825e-8dbd-42ab-9f65-aec2c5a4b6e3)

## Changes

### Jobs

The main logic change is on the `process` function of the `algoliaProductIndex` and `algoliaProductDeltaIndex` jobs. The algorithm now works as follow:
- if `color_variations` is present in the attributesToSend list:
  - if the product is a variant, we skip it. Variants records will be generated when we process the master product
  - if the product is a master, we call `generateVariantRecordsWithColorVariations()` to generate records for all its variants
  - otherwise (product set, option product), we fallback to the existing process code
- if `color_variations` is not declared, we execute the existing process code

### New attributes

- To be able to display swatches, a `color_variations` attribute can be indexed. When present in the attributes list, a `color_variations` attribute is added to each record. It's an array containing one entry per value of the `color` variation attribute. Each entry of the array has:
  - a `image_groups` attribute, with the same format than the existing `image_groups` attribute. But one additional viewtype has been added to it: `swatch`
  - a `variant_url` attribute, to make the swatches clickable

- A `master_id` attribute has also been added to the definitions, to permit records grouping.

### New helpers

- the `jobHelper` has a new method: `generateVariantRecordsWithColorVariations()`. It contains the common code shared by `algoliaProductIndex` and `algoliaProductDeltaIndex` jobs to generate, from a master product, all variant records, with the `color_variations` attribute
- new `modelHelper` which contains the existing `getImageGroups()` method that generates `image_groups` (shared with `algoliaLocalizedProduct` model) and a `getColorVariations()` method to generate the `color_variations` of a given product.

### Frontend

Both SFRA and SiteGenesis InstantSearch templates have been updated to display swatches if `color_variations` are present in the payload.

### Misc

- We were also indexing VariationGroup products. Those are not actual products and like master, shouldn't be indexed. The `productFilter` has been updated to ignore them.
- Reduce the default chunk size. When indexing variation_colors, one master product will generate many variant records. To avoid payloads growing too much, the chunk-size has been reduced.

### Tests
- new `MasterProduct` mock. It's almost the same than the existing `Product` mock, but implemented with classes and a proper constructor, permitting customization and inheritance (to create Variant mocks)
- new `Variant` mock.

## How to test :test_tube: 

- Add `color_variations` and `master_id` fields to the Additional Product Attributes
- Run the AlgoliaProductIndex_v2 jobs
  - Algolia records should now contain those 2 additional attributes
- Go to the Storefront. Color swatches should be present.
  - Hovering a swatch changes the product image
  - Clicking on a swatch brings you to the correct color variation.

---
SFCC-47
SFCC-188
SFCC-229
SFCC-238
